### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by jeff-vincent**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,11 @@
-# go-gin-redis-mongodb
\ No newline at end of file
+# go-gin-redis-mongodb
+
+![Diagram](https://raw.githubusercontent.com/jeff-vincent/go-gin-redis-mongodb/dev/images/Untitled%20Diagram-32.drawio.svg)
+
+The Web API has routes for uploading and viewing blog posts, as well as a route to get the number of times a given post has been viewed. 
+
+When a post is uploaded, it is published to the Redis channel "Upload," to which the db_worker is subscribed. When the db_worker gets the message from Redis, it inserts the post into Mongo1.
```
Comment (IamShobe) [inline README.md:7]: FYI, you can control the NL (new lines) in Markdown using "  " (double space).
So you can avoid long lines like this if you want to :).

File: redis-gin-mongo/k8s/templates/redis.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+metadata:
+  annotations:
+    velocity.tech.v1/id: redis
+  name: redis
+  labels:
+    app: redis
+spec:
+  selector:
+    matchLabels:
+      api: redis
```
Comment (IamShobe) [inline redis-gin-mongo/k8s/templates/redis.yml:13]: minor:
is there a reason not to just use "app" key label as match selector? 🤔
Not that it's really matter, cuz you configured it right

File: redis-gin-mongo/k8s/values.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+service_host: service
+service_port: "8080"
```
Comment (IamShobe) [inline redis-gin-mongo/k8s/values.yaml:2]: you basically can clean up this values file alot.
a more common practice is to use nested structs:
```
service:
   host: service
   port: "8080"
mongo1: 
   host: "mongo1"
   port: "27017"
```

it looks much more ordered like that.

File: redis-gin-mongo/k8s/values.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+mongo2_host: mongo2
+mongo2_port: "27017"
+redis_host: redis
+redis_port: "6379"
+web_api_host: web_api
+web_api_port: "8081"
+ingress_host: null
+ingress_class_name: kong
+host: "127.0.0.1"
+protocol: "http"
```
Comment (IamShobe) [inline redis-gin-mongo/k8s/values.yaml:14]: this is very confusing value to be honest - it is difficult to understand what it is relates to in the chart without reading the actual chart.
a much better approach will be using something like:
```
webApi:
   envVars:
     PROTOCOL: "http"
     # put all the rest env vars here
```

then in the chart you can expand it:
```
{{- $key, $value := range $.Values.webAPI.envVars -}}
{{ $key }}: {{ $value | quote }}
{{- end -}}
```

then the values file is much more readable :)

Comment (IamShobe) [reply redis-gin-mongo/k8s/values.yaml:14]: I would also extract even more fields from the chart itself - like the annotations segment, and image and tag itself.
the reason:
- for the annotations - I think it the best to have a single velocity values file that contains all "logic" of velocity - including defining dependencies and setting up template values.
- for the image and tag - basically most of the time those things change quite often and the last thing you really want to do is to update the version tag of the chart (which in turn means you basically need to publish the chart into a chart museum and make everyone update, instead of just providing a new values file).
- you might also consider extracting replicas count

File: redis-gin-mongo/k8s/templates/db_worker2.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      labels:
+        app: db-worker2
+        api: db-worker2
+    spec:
+      containers:
+        - name: db-worker2
+          image: jdvincent/gin_redis_db_worker2:latest
+          env:
+            - name: REDIS_HOST
+              value: {{ .Values.redis_host | toJson  }}
```
Comment (IamShobe) [inline redis-gin-mongo/k8s/templates/db_worker2.yml:27]: prefer using `| quote` instead :) as it's clearer what is the goal there

File: redis-gin-mongo/src/analytics_service/go.mod [hunk 6]
```diff
@@ -0,0 +1,38 @@
+module example.com/m/v2
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/go.mod:1]: you can actually update this module to point to your repository.
This will also allow you to properly have imports between different go modules

File: redis-gin-mongo/src/analytics_service/main.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+var MONGO2_HOST = os.Getenv("MONGO2_HOST")
+var MONGO2_PORT = os.Getenv("MONGO2_PORT")
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: minor suggestion: you can group those using:
```
var (
  MONGO2_HOST = os.Getenv("MONGO2_HOST")
  MONGO2_PORT = os.Getenv("MONGO2_PORT")
)
```

File: redis-gin-mongo/src/analytics_service/main.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+func getDoc(client mongo.Client, title string) bson.D {
+	coll := client.Database("blog").Collection("views")
+	var result bson.D
+	err := coll.FindOne(context.TODO(), bson.D{{"title", title}}).Decode(&result)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: Umm here the context here isn't being handled correctly:
you basically should propagate the framework's (gin) context to here.
the reason is that the framework context might contain stop signal as well / or some other mechanisms for timeouts.
If you don't propagate the ctx then you basically not allowing this routine to gracefully stop (even if you use ctrl + c for example while in the query).

File: redis-gin-mongo/src/analytics_service/main.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	return result
+}
+
+func getAllDocs(client mongo.Client) []bson.M {
+	coll := client.Database("blog").Collection("views")
+	cursor, err := coll.Find(context.TODO(), bson.D{})
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go:39]: should propagate the ctx here as well :)

File: redis-gin-mongo/src/analytics_service/main.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+var MONGO2_HOST = os.Getenv("MONGO2_HOST")
+var MONGO2_PORT = os.Getenv("MONGO2_PORT")
+
+func getDoc(client mongo.Client, title string) bson.D {
+	coll := client.Database("blog").Collection("views")
+	var result bson.D
+	err := coll.FindOne(context.TODO(), bson.D{{"title", title}}).Decode(&result)
+	if err != nil {
+		log.Fatal(err)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: well I would expect the error should be handled differently :) 
few things here:
- basically in go it is really common practice to propagate the error to the calling routine (similar to how exception propagates up the stack). then you will see most of the times something like:
```
if err := someHandler(); err != nil {
 return err
}
```
or for tuple return value:
```
value, err := someHandler()
if err != nil {
   return nil, err
}
```
Then the calling parent can either continue to propagate the error:
```
err := callChild()
if err != nil {
   return errors.Wrap(err, "child faild")
}
```
or just handle it in a graceful way:
```
err := callChild()
if err != nil {
   // we failed and we can run something else instead
   if isNotFoundErr(err) {
     createANewInstance()
   }
}
```

- basically when you choose something to be a Fatal - you basically want to panic right away - fatal means there is no recovery from this, and the app "should crash" (and it does it by calling os.Exit(1)). Most of the time you dont really want to panic and basically exit gracefully when possible - therefore I think it is better to just log and handler gracefully with 404 message maybe :)

File: redis-gin-mongo/src/analytics_service/main.go [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	return result
+}
+
+func getAllDocs(client mongo.Client) []bson.M {
+	coll := client.Database("blog").Collection("views")
+	cursor, err := coll.Find(context.TODO(), bson.D{})
+	if err != nil {
+		log.Fatal(err)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: same as before :)
better return err here and an empty slice

File: redis-gin-mongo/src/analytics_service/main.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	var results []bson.M
+	if err = cursor.All(context.TODO(), &results); err != nil {
+		panic(err)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go:45]: same as before this time error code might be like 500

File: redis-gin-mongo/src/analytics_service/main.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+	var result bson.D
+	err := coll.FindOne(context.TODO(), bson.D{{"title", title}}).Decode(&result)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return result
+}
+
+func getAllDocs(client mongo.Client) []bson.M {
+	coll := client.Database("blog").Collection("views")
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: better declare those as constants at the start of this module:
```
const (
  databaseName = "blog"
  collectionName = "views"
)
```

File: redis-gin-mongo/src/analytics_service/main.go [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	var results []bson.M
+	if err = cursor.All(context.TODO(), &results); err != nil {
+		panic(err)
+	}
+	return results
+}
+
+func main() {
+	mongo_uri := fmt.Sprintf("mongodb://%s:%s", MONGO2_HOST, MONGO2_PORT)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: this can be moved to `var` declaration at the start of the module

File: redis-gin-mongo/src/analytics_service/main.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+	client, err := mongo.NewClient(options.Client().ApplyURI(mongo_uri))
+	if err != nil {
+		log.Fatal(err)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: here is a good example for "panic"/"fatal" as you defined it - the reason is that it is located at the top most level (main function) then basically you are responsible on the process actions - like termination with an exit code of your choice.
There is a single place of termination and it is expected to find terminations here.

File: redis-gin-mongo/src/analytics_service/main.go [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: missing defer of cancel
you basically want to keep it 
```
defer cancel()
```
not using it is running the risk of a memory leak actually

File: redis-gin-mongo/src/analytics_service/main.go [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+	err = client.Connect(ctx)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer client.Disconnect(ctx)
+	r := gin.Default()
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: prefer using more readable way here, the "Python" way :).
I know that golang is a little harsh about that, but I do really value readability over all.
```
router := gin.Default()
```
is really much more better - and allows others that are not familiar with gin framework to understand what `gin.Default()` generates

File: redis-gin-mongo/src/analytics_service/main.go [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+
+	r.GET("/get-doc", func(c *gin.Context) {
+		title := c.Query("title")
+		result := getDoc(*client, title)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: consider changing `getDoc` to receive the client as a pointer as well.

```
func getDoc(ctx *gin.Context, client *mongo.Client, title string) bson.D {
```

File: redis-gin-mongo/src/analytics_service/main.go [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/gin-gonic/gin"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+var MONGO2_HOST = os.Getenv("MONGO2_HOST")
+var MONGO2_PORT = os.Getenv("MONGO2_PORT")
+
+func getDoc(client mongo.Client, title string) bson.D {
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: consider rename to:
```
getAnalyticsDataByTitle
```

`Document` terminology is actually very "low level" detail - the fact you are using Mongo shouldn't effect the logic itself.
also you usually prefer returning a domain model instead of plain bson.D.
this will allow you to keep the same signature (most of the time) while changing inside implementations.

usually you can have some kind of "repository" / CRUD / DAO:
```
type AnalyticsDataMongoDAO struct {
  client mongo.Client
}

func (dao *AnalyticsDataMongoDAO) getAnalyticsDataByTitle(title string) (*AnalyticsData, error) {

}
```

then you basically can have an interface of this DAO:
```
type AnalyticsDataDAO interface {
   getAnalyticsDataByTitle(title string) (*AnalyticsData, error)
}
```

and using dependency injection you can request where you use it for a DAO that is capable of querying AnalyticsData.
Then if there's a day that you want to migrate to different DB:
```
type AnalyticsDataPSQLDAO struct {
  client psql.Client
}

func (dao *AnalyticsDataPSQLDAO) getAnalyticsDataByTitle(title string) (*AnalyticsData, error) {

}
```

you basically have the exact same interface - and need to change nothing else then the actual implementation of the handler.

this is becoming much more useful in a much larger scale applications, where it is becoming really scary to make huge changes - so you want to use dependency injections to allow better testing capabilities and just implementing behaviours :)

File: redis-gin-mongo/src/analytics_service/main.go [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+	coll := client.Database("blog").Collection("views")
+	var result bson.D
+	err := coll.FindOne(context.TODO(), bson.D{{"title", title}}).Decode(&result)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return result
+}
+
+func getAllDocs(client mongo.Client) []bson.M {
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: consider rename to:
```
getAllBlogViews
```

File: redis-gin-mongo/src/analytics_service/main.go (hunk 19 above)
Comment (IamShobe) [reply redis-gin-mongo/src/analytics_service/main.go]: you can also use an actual struct model instead of `bson.D` this will actually be much better because it will be clearer what is returning from here - without reading the actual insert query of the other micro services. https://www.mongodb.com/docs/drivers/go/current/fundamentals/bson/

File: redis-gin-mongo/src/analytics_service/main.go [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+		c.JSON(http.StatusOK, gin.H{
+			"Data": result,
+		})
+	})
+
+	r.GET("/get-all-docs", func(c *gin.Context) {
+		result := getAllDocs(*client)
+		c.JSON(http.StatusOK, gin.H{
+			"Data": result,
+		})
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_service/main.go]: here you can use a struct instead that uses json struct tags:
```
type GetAllDocsResponse struct {
   Data []View `json:"data"`
}
```

you should prefer this for type safety and readability :)

File: redis-gin-mongo/src/analytics_worker/go.mod [hunk 22]
```diff
@@ -0,0 +1,24 @@
+module example.com/m/v2
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/go.mod:1]: here as we said before you can use actual github path

File: redis-gin-mongo/src/analytics_worker/main.go [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+var MONGO2_PORT = os.Getenv("MONGO2_PORT")
+var REDIS_HOST = os.Getenv("REDIS_HOST")
+var REDIS_PORT = os.Getenv("REDIS_PORT")
+
+type AnalyticsData struct {
+	Title string
+	Views int
+}
+
+func getDoc(title string, mongo2 *mongo.Client) AnalyticsData {
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/main.go]: consider:
- keeping the same ordering across all modules: if you chose client to be first - then it should be first in the params list.
- rename to `getAnalyticsDataByTitle`

File: redis-gin-mongo/src/analytics_worker/main.go [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+	coll := mongo2.Database("blog").Collection("views")
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/main.go]: actually I said before this can be moved to consts - but because you are sharing those across 2 modules - it is much better implementing a DAO that is shared between of them - then there is a single module that is responsible for every operation of this collection.
Less error prune and much more safe :)

File: redis-gin-mongo/src/analytics_worker/main.go [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+	var result AnalyticsData
+	err := coll.FindOne(context.TODO(), bson.D{{"title", title}}).Decode(&result)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/main.go:39]: as before, ctx propagation, and use struct instead for type safety

File: redis-gin-mongo/src/analytics_worker/main.go [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+	if err != nil {
+		fmt.Println(err)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/main.go]: return the error in this case and return 404:
```


if err == mongo.ErrNoDocuments {
// This error means your query did not match any documents.
return NewViewsNotFoundErr()
}


```

(you should return an error that is not mongo specific to allow other databases to return the same error and not change the rest of the code).

File: redis-gin-mongo/src/analytics_worker/main.go [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+	}
+	return result
+}
+
+func insertDoc(title string, mongo2 *mongo.Client) *mongo.InsertOneResult {
+	coll := mongo2.Database("blog").Collection("views")
+	data := AnalyticsData{Title: title, Views: 0}
+	res, err := coll.InsertOne(context.TODO(), data)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/main.go]: that is great you used here actual struct :) 
dont forget the ctx here as well

File: redis-gin-mongo/src/analytics_worker/main.go [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func updateAnalytics(title string, mongo2 *mongo.Client) {
+	existingDoc := getDoc(title, mongo2)
+	if existingDoc.Title == "" {
+		insertDoc(title, mongo2)
+	} else {
+		views := existingDoc.Views + 1
+		coll := mongo2.Database("blog").Collection("views")
+		result, err := coll.UpdateOne(
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/main.go]: for this you can actually use `ReplaceOne` and then use the struct itself after update :)

File: redis-gin-mongo/src/analytics_worker/main.go [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+			context.TODO(),
+			bson.M{"title": existingDoc.Title},
+			bson.D{
+				{"$set", bson.D{{"views", views}}},
+			},
+		)
+		if err != nil {
+			log.Fatal(err)
+		}
+		fmt.Println(result)
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/main.go]: prefer using the log if you already use it elsewhere

File: redis-gin-mongo/src/analytics_worker/main.go [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+	coll := mongo2.Database("blog").Collection("views")
+	var result AnalyticsData
+	err := coll.FindOne(context.TODO(), bson.D{{"title", title}}).Decode(&result)
+	if err != nil {
+		fmt.Println(err)
+	}
+	return result
+}
+
+func insertDoc(title string, mongo2 *mongo.Client) *mongo.InsertOneResult {
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/main.go]: prefer: 
`createAnalyticsData`

File: redis-gin-mongo/src/analytics_worker/main.go [hunk 31]
```diff
@@ -0,0 +1,94 @@
...
+	"time"
+
+	"github.com/go-redis/redis/v8"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+var MONGO2_HOST = os.Getenv("MONGO2_HOST")
+var MONGO2_PORT = os.Getenv("MONGO2_PORT")
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/main.go]: prefer reusing those from a shared place

File: redis-gin-mongo/src/analytics_worker/main.go [hunk 32]
```diff
@@ -0,0 +1,94 @@
...
+	}
+	defer mongo2.Disconnect(ctx)
+
+	redis_uri := fmt.Sprintf("redis://%s:%s/0", REDIS_HOST, REDIS_PORT)
+	opt, err := redis.ParseURL(redis_uri)
+	if err != nil {
+		panic(err)
+	}
+	rdb := redis.NewClient(opt)
+	pubsub := rdb.Subscribe(ctx, "Analytics")
```
Comment (IamShobe) [inline redis-gin-mongo/src/analytics_worker/main.go]: const it

File: redis-gin-mongo/src/blog_service/main.go [hunk 33]
```diff
@@ -0,0 +1,73 @@
...
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+var MONGO1_HOST = os.Getenv("MONGO1_HOST")
+var MONGO1_PORT = os.Getenv("MONGO1_PORT")
+var REDIS_HOST = os.Getenv("REDIS_HOST")
+var REDIS_PORT = os.Getenv("REDIS_PORT")
+
+func getDoc(client mongo.Client, title string) bson.D {
+	coll := client.Database("blog").Collection("posts")
```
Comment (IamShobe) [inline redis-gin-mongo/src/blog_service/main.go]: const those params as well.
use structs and implement DAO if possible :)

File: redis-gin-mongo/src/blog_service/main.go [hunk 34]
```diff
@@ -0,0 +1,73 @@
...
+	redis_uri := fmt.Sprintf("redis://%s:%s/0", REDIS_HOST, REDIS_PORT)
+	opt, err := redis.ParseURL(redis_uri)
+	if err != nil {
+		panic(err)
+	}
+	rdb := redis.NewClient(opt)
+	ctx := context.Background()
+	err = rdb.Publish(ctx, "Analytics", payload).Err()
+	if err != nil {
+		panic(err)
```
Comment (IamShobe) [inline redis-gin-mongo/src/blog_service/main.go]: prefer not panicing but returning 500 error / 503 error after propagation

File: redis-gin-mongo/src/blog_service/main.go [hunk 35]
```diff
@@ -0,0 +1,73 @@
...
+func Publish(payload string) {
+	redis_uri := fmt.Sprintf("redis://%s:%s/0", REDIS_HOST, REDIS_PORT)
+	opt, err := redis.ParseURL(redis_uri)
+	if err != nil {
+		panic(err)
+	}
+	rdb := redis.NewClient(opt)
+	ctx := context.Background()
+	err = rdb.Publish(ctx, "Analytics", payload).Err()
+	if err != nil {
```
Comment (IamShobe) [inline redis-gin-mongo/src/blog_service/main.go]: you can use short hand:
```
if err = rdb.Publish(ctx, "Analytics", payload).Err(); err != nil {

}
```

File: redis-gin-mongo/src/blog_service/main.go [hunk 36]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	defer client.Disconnect(ctx)
+	r := gin.Default()
+
+	r.GET("/get-doc", func(c *gin.Context) {
+		title := c.Query("title")
+		result := getDoc(*client, title)
+
+		c.JSON(http.StatusOK, gin.H{
+			"Data": result,
```
Comment (IamShobe) [inline redis-gin-mongo/src/blog_service/main.go:88]: return type as struct :)

File: redis-gin-mongo/src/blog_service/main.go [hunk 37]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
+	err = client.Connect(ctx)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer client.Disconnect(ctx)
+	r := gin.Default()
+
+	r.GET("/get-doc", func(c *gin.Context) {
```
Comment (IamShobe) [inline redis-gin-mongo/src/blog_service/main.go]: prefer using more meaningful names for routes here and everywhere else, also we can follow the convention for REST api:
```
/posts/
```
to receive all posts
```
/posts/:id
```
to receive a single one (in your case id is title)
so prefer changing the query param to path params :)

File: redis-gin-mongo/src/db_worker/main.go [hunk 38]
```diff
@@ -0,0 +1,66 @@
...
+	pubsub := rdb.Subscribe(ctx, "Upload")
+	defer pubsub.Close()
+	ch := pubsub.Channel()
+	for msg := range ch {
+		post := BlogPost{}
+		if err := json.Unmarshal([]byte(msg.Payload), &post); err != nil {
+			panic(err)
+		}
+		insertDoc(client, post)
+		fmt.Println(post)
```
Comment (IamShobe) [inline redis-gin-mongo/src/db_worker/main.go]: use log instead :)

File: redis-gin-mongo/src/db_worker/main.go [hunk 39]
```diff
@@ -0,0 +1,66 @@
...
+		panic(err)
+	}
+	rdb := redis.NewClient(opt)
+	pubsub := rdb.Subscribe(ctx, "Upload")
+	defer pubsub.Close()
+	ch := pubsub.Channel()
+	for msg := range ch {
+		post := BlogPost{}
+		if err := json.Unmarshal([]byte(msg.Payload), &post); err != nil {
+			panic(err)
```
Comment (IamShobe) [inline redis-gin-mongo/src/db_worker/main.go]: well this might be problematic you dont handle it garcefully.
Lets say we got a real error here - for example an attacker sends a payload that is not json unmarshal-able - you basically allow for DOS (denial of service) here and the process will keep crashing.
because this is a worker - then you can't really return 400 error - then in this case I will just log an error, and continue to next iteration.
(usually because this is working asyncly a client will sample the response every few seconds and will receive the error message).

File: redis-gin-mongo/k8s/templates/db_worker1.yml [hunk 40]
```diff
@@ -0,0 +1,33 @@
...
+    velocity.tech.v1/id: db-worker1
+    velocity.tech.v1/dependsOn: redis, mongo1
+  name: db-worker1
+  labels:
+    app: db-worker1
+spec:
+  selector:
+    matchLabels:
+      api: db-worker1
+  replicas: 3
```
Comment (IamShobe) [inline redis-gin-mongo/k8s/templates/db_worker1.yml:15]: pay attention that redis channels is working as "pub/sub" meaning that all workers will receive the same message and operate the same work.

File: redis-gin-mongo/src/web_api/main.go [hunk 41]
```diff
@@ -0,0 +1,124 @@
...
+var REDIS_PORT = os.Getenv("REDIS_PORT")
+var ANALYTICS_SERVICE_HOST = os.Getenv("ANALYTICS_SERVICE_HOST")
+var ANALYTICS_SERVICE_PORT = os.Getenv("ANALYTICS_SERVICE_PORT")
+var BLOG_SERVICE_HOST = os.Getenv("BLOG_SERVICE_HOST")
+var BLOG_SERVICE_PORT = os.Getenv("BLOG_SERVICE_PORT")
+
+func getDoc(c *gin.Context) {
+	title := c.Query("title")
+	address := fmt.Sprintf("http://%s:%s/get-doc?title=%s", BLOG_SERVICE_HOST, BLOG_SERVICE_PORT, title)
+	resp, _ := http.Get(address)
```
Comment (IamShobe) [inline redis-gin-mongo/src/web_api/main.go:44]: err is being ignored here (possible 503 for example / 404)

File: redis-gin-mongo/src/web_api/main.go [hunk 42]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+var REDIS_HOST = os.Getenv("REDIS_HOST")
+var REDIS_PORT = os.Getenv("REDIS_PORT")
+var ANALYTICS_SERVICE_HOST = os.Getenv("ANALYTICS_SERVICE_HOST")
+var ANALYTICS_SERVICE_PORT = os.Getenv("ANALYTICS_SERVICE_PORT")
+var BLOG_SERVICE_HOST = os.Getenv("BLOG_SERVICE_HOST")
+var BLOG_SERVICE_PORT = os.Getenv("BLOG_SERVICE_PORT")
+
+func getDoc(c *gin.Context) {
```
Comment (IamShobe) [inline redis-gin-mongo/src/web_api/main.go]: consider renaming :)

File: redis-gin-mongo/src/web_api/main.go [hunk 43]
```diff
@@ -0,0 +1,124 @@
...
+	decoder := json.NewDecoder(resp.Body)
+	err := decoder.Decode(val)
+
+	if err != nil {
+		fmt.Println(err)
+	}
+	c.JSON(http.StatusOK, val)
+}
+
+func getAllDocs(c *gin.Context) {
```
Comment (IamShobe) [inline redis-gin-mongo/src/web_api/main.go]: consider renaming :)

File: redis-gin-mongo/src/web_api/main.go [hunk 44]
```diff
@@ -0,0 +1,124 @@
...
+	if err != nil {
+		panic(err)
+	}
+	rdb := redis.NewClient(opt)
+
+	r := gin.Default()
+	r.LoadHTMLGlob("templates/*.html")
+
+	r.GET("/", index)
+	r.POST("/insert-doc", func(c *gin.Context) {
```
Comment (IamShobe) [inline redis-gin-mongo/src/web_api/main.go]: REST wise a better way is to just POST to `/posts` :) 
`GET /posts` = get all posts
`POST /posts` = post a new post
`PUT /posts/:id` = edit post `:id`
`DELETE /posts/:id` = delete post `:id`
`GET /posts/:id` = get post `:id`

File: redis-gin-mongo/src/web_api/main.go [hunk 45]
```diff
@@ -0,0 +1,124 @@
...
+		title := c.PostForm("title")
+		author := c.PostForm("author")
+		body := c.PostForm("body")
+		new_post := BlogPost{Title: title, Author: author, Body: body}
+		payload, err := json.Marshal(new_post)
+		if err != nil {
+			fmt.Println(err)
```
Comment (IamShobe) [inline redis-gin-mongo/src/web_api/main.go]: log instead :) 
and return 400
not sure if err is even possible here actually 🤔

File: redis-gin-mongo/src/web_api/main.go [hunk 46]
```diff
@@ -0,0 +1,124 @@
...
+		err = rdb.Publish(ctx, "Upload", payload).Err()
+		if err != nil {
+			panic(err)
+		}
+		newPost(c, title, author, body)
+	})
+	r.GET("/get-doc", getDoc)
+	r.GET("/get-all-docs", getAllDocs)
+	r.GET("/get-post-views", getPostViews)
+	r.Run(":8081")
```
Comment (IamShobe) [inline redis-gin-mongo/src/web_api/main.go]: prefer receiving listening port from env var

File: redis-gin-mongo/src/web_api/main.go [hunk 47]
```diff
@@ -0,0 +1,124 @@
...
+		body := c.PostForm("body")
+		new_post := BlogPost{Title: title, Author: author, Body: body}
+		payload, err := json.Marshal(new_post)
+		if err != nil {
+			fmt.Println(err)
+		}
+		ctx := context.Background()
+		err = rdb.Publish(ctx, "Upload", payload).Err()
+		if err != nil {
+			panic(err)
```
Comment (IamShobe) [inline redis-gin-mongo/src/web_api/main.go]: return 503 instead

**Review by IamShobe — COMMENTED** [review]
WOW, man this is so great!
I really like your work!!!

I tried to go as strict as possible for the learning experience, but don't feel obligated for any of the comments.
I commented only once on most repeating things so for every comment make sure to go over all the code to cover everything related :)

Keep up the great work !!! 💪
